### Preprocessing and input signature generation
Basically, this is how author has processed the data. Just follow this pipeline as a guideline. It is okay to proceed other tools. Just make sure to match the format of output as what this package has used. 

If anyone who adopts one of the processes in this pipeline, please cite the original package (mandatory).

We used lung cancer data from (Maynard, A. et al. Therapy-Induced Evolution of Human Lung Cancer Revealed by Single-Cell RNA Sequencing. Cell 182, 1232-1251 e1222, doi:10.1016/j.cell.2020.07.017 (2020)) paper. Please download the Seurat object if needed.

## 1. Variant calling
For variant calling from fastq files, we used CTAT (https://github.com/NCIP/Trinity_CTAT).
(citation: Fangal, Vrushali Dipak. 2020. CTAT Mutations: A Machine Learning Based RNA-Seq Variant Calling Pipeline Incorporating Variant Annotation, Prioritization, and Visualization. Master's thesis, Harvard Extension School)

* Script:inst/1_variant_calling_paired.sh, inst/1_variant_calling_single.sh

``` bash
./ 1_variant_calling_paired.sh SRR13157248 4
#fastq_files: SRR13157248_1.fastq, SRR13157248_2.fastq
#4 cores
```
* Output: ####.filtered.vcf.gz for the further process (ex: SRR13157248.filtered.vcf.gz).

## 2. Variant filtering
We further filtered out non-COSMIC SNPs found in dbSNP and RNA-editing sites provided from CTAT with filtered.vcf files.

* Script: inst/ 2_vcf_processing
* Data: data/vcf_list

``` r
#vcf_list: list of vcf files of each single-cell
#there should be  "../vcf_new/" directory
library(vcfR)
library(ape)

a<-readLines("vcf_list")

for (j in c(1: length(a))){
	tmp_vcf <- a[j]
	vcf<-read.vcfR(tmp_vcf)

	k<-sapply(vcf@fix[,8],function(x) (!grepl("COSMIC",x) & grepl("RS=",x)|grepl("RNAEDIT",x)))
	k1 <- which(k)
	vcf@fix[k1,7] <- "notpass"

	vcf1<-vcf[vcf@fix[,7] =="PASS",]

	new_name <- paste("../new_vcf/", tmp_vcf, sep = "")
	#there should be  "../vcf_new/" directory

	write.vcf(vcf1, file=new_name, mask=TRUE)

}

```

## 3. VCF merging
Merging vcf files with same sample or origin
(https://samtools.github.io/bcftools/bcftools.html#merge)

* Script: inst/ 2_vcf_processing
* Data: same formet as vcf_list (but only with the same sample)

``` bash
module load bcftools
bcftools merge -l patient_vcf_list --no-index -o lung_patient1.vcf.gz -0
```

## 4. Clonotyping
We performed DENDRO to infer clonotype for each cancer cell (Basically, any other tool is okay to infer clonotype).Finally, we included clonotype information into Seurat object.
(https://github.com/zhouzilu/DENDRO)
(citation: Zhou, Z., Xu, B., Minn, A. & Zhang, N. R. DENDRO: genetic heterogeneity profiling and subclone detection by single-cell RNA sequencing. Genome Biol 21, 10, doi:10.1186/s13059-019-1922-x (2020))

* Script: inst/3_dendro_pipeline1.R, inst/3_dendro_pipeline2.R, inst/dendro_function.R
* Data: data/pat_optk, Seurat object (name: d2, not provided)

``` r
library(DENDRO)
library(data.table)
library(tidyr)
library(dplyr)

source("inst/dendro_function.R")
name <- "merged_patient"
#ex: AZ003.vcf.gz -> AZ003

dendro_process1(name)
```

* Output: patient_demo_qc (ex: AZ003_demo_qc)
After making “data/pat_optk” file, (Please see the format of the file) ...

``` r
library(DENDRO)
source("inst/dendro_function.R")
pat_data <- read.table("data/pat_optk", sep = "\t")
#load("data/lung_seurat")
#not provided
#seurat object; name: d2

d2$cancer_clone <- NA
b1<-names(d2$orig.ident)

for (j in c(1: dim(pat_data)[1])){
        print(rownames(pat_data)[j])
        tmp_label <- dendro_process2(rownames(pat_data)[j],pat_data$optk[j])
        a1<-names(tmp_label)
        a2<-match(a1,b1)
        d2$cancer_clone[a2] <- tmp_label
}


save(d2, file = "lung_seurat_mut")
```

* Output: patient_final_dendro (ex: AZ003_final_dendro)

## 5. Subclonal evolution tree generation
We generated subclonal evolution tree from genotype information of each clonotype generated by DENDRO with RobustClone.
(https://github.com/ucasdp/RobustClone)
(citation: Chen, Z., Gong, F., Wan, L. & Ma, L. RobustClone: a robust PCA method for tumor clone and evolution inference from single-cell sequencing data. Bioinformatics 36, 3299-3306, doi:10.1093/bioinformatics/btaa172 (2020))

* Script: 4_robust_clone, robust_clone_function.R
* Data: patinet_final_dendro (output from Dendro), Seurat object (name: d2, not provided)

``` r
# Robust clone for generating clonal minimum spanning tree
source('inst/robust_clone_function.R')

load("lung_seurat_mut")
#data not provided
#seurat object; name: d2

pat_list <- unique(d2$pat)
for (j in c(1:length(pat_list))){
	pat_tmp <- pat_list[j]
	print(pat_tmp)
	load(paste("data/", pat_tmp, "_final_dendro", sep = ""))
p<-subset(d2, subset = pat == pat_tmp)
	p1 <- na.omit(p$cancer_clone)
	clone_gety<-t(demo_cluster$Z)
	robust_clone <- c()
	for(i in 1:length(unique(p1))){
		robust_clone <- c(robust_clone,list(which(p1==i)))
	}

	clone_path <- plot_MST(clone_gety, robust_clone, 'SNV', pat_tmp)
	write.table(clone_path, paste(pat_tmp, "_clone_path", sep = ""), sep = "\t", quote=F)

}
```

* Output: patient_clone_path (ex:AZ003_clone_path)
[see data/robustclone directory]

* Usage
``` r
clone_path <- read.table("data/robustclone/AZ003_clone_path")
```
![image](https://user-images.githubusercontent.com/54588204/203587438-606edb47-0cae-4c8f-85a1-3bbc3493d8a5.png)

## 6. Mutation list object generation
Make mutation profiles for each clonotype as list object in R

* Script: 5_make_clone_mut_profile

``` r
a<-read.table("data/meta_info_lung", sep = "\t")
sample_list <- as.character(a[,"sample_name"])
sample_list_uniq <- unique(sample_list)
pat_list <- as.character(a[,"pat_collapse"])
sample_pat <- table(sample_list, pat_list)

make_mut_list(sample_list_uniq, sample_pat, dendro_directory)
```

* Output: data/clone_mutation_info/sample_clone_mut_list (sample: sample name)

* Usage
``` r
load("data/clone_mutation_info/LT_S21_clone_mut_list")
```
![image](https://user-images.githubusercontent.com/54588204/203589014-a640de18-5531-4722-85f5-12cd51fab330.png)

 

## 7. Clonal (cluster) abundance
Obtaining relative abundance of each clone or cluster

* Script: inst/6_abundance
* Data: Seurat object with only cancer cells (name: d3, not provided)

``` r
load("lung_cancer_seurat_mut")
#not provided

###clone abundance
a<-table(d3$sample_name,d3$cancer_clone)
colnames(a) <- paste("clone", colnames(a), sep = "_")
write.csv(a, "lung_clone_cnt.csv", quote = F)

a1<-a/rowSums(a)
write.csv(a1, "lung_clone_ratio.csv", quote = F)

###cluster abundance
a<-table(d3$sample_name,d3$seurat_clusters)
colnames(a) <- paste("cluster", colnames(a), sep = "_")
write.csv(a, "lung_cluster_cnt.csv", quote = F)

a1<-a/rowSums(a)
write.csv(a1, "lung_cluster_ratio.csv", quote = F)
```
* Usage
``` r
a <- read.csv("data/lung_clone_ratio.csv")
```
 
![image](https://user-images.githubusercontent.com/54588204/203589051-cab15e5f-b7f1-4f47-86aa-893daf3f2142.png)

## 8. Inferring cell-cell interaction
We used CellChat package for inferring cell-cell interaction. 
(https://github.com/sqjin/CellChat)
(citation: Jin, S. et al. Inference and analysis of cell-cell communication using CellChat. Nat Commun 12, 1088, doi:10.1038/s41467-021-21246-9 (2021).)
Cancer clustering (by gene expression) should be done beforehand. Basically, users can use their own package with the same format of output.

* Script: inst/8_cellchat

``` r
library(CellChat)
library(Seurat)
CellChatDB <- CellChatDB.human
load("lung_seurat_mut_final")
#seurat object, name: d2

cellchat_run(d2, "lung", "celltype_cancer")

##
cellchat_rl("lung")

```

* Output: data/cellchat/lung_source_gn, data/cellchat/lung_target_gn
source_gn: source from cancer cell
target_gn: target from cancer cell
List object name: res

* Usage
``` r
load("data/cellchat/lung_source_gn")
names(res)
# target celltypes
res$Dendritic
#mutation profiles of each cancer cluster (by gene expression)
```
![image](https://user-images.githubusercontent.com/54588204/203589096-359b14fb-a950-4dfc-b158-003ae42161e3.png)
 

## 9. Preparation of gmt format for geneset
Please prepare geneset file with gmt format

* Usage
``` r
library(GSEABase)
gmt <- getGmt("data/cancersea_gmt")
```

# Please don’t forget to cite the original paper (or relevant citation) when using their packages

